### ISSUE TITLE
Tab completion issue

### ISSUE THREAD

**Issue by Sukram28** [ProtocolSupport/ProtocolSupport#1171]
**Report**
I used the PlayerCommandSendEvent to clear all command suggestions and to add own one:

`@EventHandler
public void onTab(PlayerCommandSendEvent e) {
	e.getCommands().clear(); e.getCommands().add("help");
}`

The Problem: 1.4.7-1.12.2 Clients can see all hidden command suggestions.
I've tried many different things, like canceling the PacketType.Play.Client.TAB_COMPLETE PacketEvent or canceling the TabCompleteEvent. Nothing works. I don't want to remove the complete Tab suggestions for 1.13-1.14 Clients. Is there a way to fix this, or a way to simply disable Tab completion and Commandsuggestion for 1.4.7-1.12.2 Clients?

**How to recreate**
1) Use a 1.13-1.14 Server with ProtocolSupport. 
2) Add a plugin which implements a Listener with `@EventHandler
public void onTab(PlayerCommandSendEvent e) {
	e.getCommands().clear();
}`
3) Join the Server with a 1.12 Client and type in chat / + [TAB] 

**ProtocolSupport version (/ps buildinfo)**
protocolsupport.ProtocolSupport$BuildInfo(buildtime: 2019.04.09 at 18:42:11 MSK, buildhost: build.true-games.org, buildnumber: 377)

**Server version**
paper-1.13.2, paper1.14.4, spigot-1.13.2, spigot 1.14.4 latest versions

**Comment by Shevchik** [comment]
That's how it works on those versions. The never had client sided command completions.

**Comment by Sukram28** [comment]
> That's how it works on those versions.
Can't agree with this.

Tab completion could be easly removed with for example the spigot option:
`  tab-complete: -1` 

or by using the following code:
`ProtocolLibrary.getProtocolManager().addPacketListener(new PacketAdapter(plugin, ListenerPriority.NORMAL,
				new PacketType[] { PacketType.Play.Client.TAB_COMPLETE }) {
			public void onPacketReceiving(PacketEvent e) {
				try {
					if (e.getPacketType() == PacketType.Play.Client.TAB_COMPLETE) {
						if (e.getPlayer().hasPermission(NoTab.plugin.tabperm)) {
							return;
						}
						PacketContainer packet = e.getPacket();
						String message = ((String) packet.getSpecificModifier(String.class).read(0)).toLowerCase();
						if ((message.startsWith("/")) && (!message.contains(" "))) {
							e.setCancelled(true);
							if (!NoTab.tabarray.contains(e.getPlayer())) {
								NoTab.tabarray.add(e.getPlayer());
								NoTab.removePlayer(e.getPlayer());
							}
						}
					}
				} catch (Exception ex) {
					NoTab.plugin.getLogger().severe("[NoTab] Error accesing to packet");
				}
			}
		});`

There is no way to disable command suggestions, which is a huge safety issue

**Comment by Shevchik** [comment]
It's not a protocolsupport job to do this. < 1.13 clients request tab completions for server (and you can do it on 1.13+ by modifying client too). It's server admin job to do this.
Don't see any safety issues with this.

**Comment by Sukram28** [comment]
> It's not a protocolsupport job to do this. It's server admin job to modify this.

I tried to modify it but it doesn't work for Clients 1.12-. The tab-complete option of Spigot also only works for Clients with the same Protocol version of the server. If there would be some was to simply remove the command suggestions for 1.12- Clients, I would do it but nothing seems to be effective. In 1.12 I could disable all command suggestions with the code above or by changing tab-complete to -1. 

> < 1.13 clients request tab completions for server (and you can do it on 1.13+ by modifying client too

I'm not sure if I got this right, but as I said above it's not possible to modify/remove somehow the tab complete funtion for 1.12- Clients on 1.13+ Servers. 

> Don't see any safety issues with this.

I think it's a huge safety issue. Many Plugin developers don't know how to hide command suggestions. 1.12- Clients just need to type /[Tab] and get an entire list of all plugins that the server uses. Commands which they shouldn't see and probably can't even execute. They getting a list of AntiCheat Plugins and can simply use specific hackclients which can handle those anticheat plugins and even worse: With all plugin information they can copy Server or look for specific things which aren't blocked, like some dupe exploits which can be prevented by some plugins or Anti AFK-Detectors and so on. 

At least it should be possible to prevent command suggestions for 1.12- Clients if the spigot option:
`tab-complete: -1` 
is set.
